### ISSUE TITLE
Simultaneous AtlasDB client start up failure - internal table creation (e.g. _metadata / _locks) continues to suffer from CASSANDRA-9933.

### ISSUE THREAD

**Issue by jboreiko** [palantir/atlasdb#444]
**BLUF: Testing of an internal product has determined that the simultaneous start of multiple AtlasDB clients can indeed cause a crash as a result of a race condition in the creation of two internal tables.**

The crash is linked to the propagation of a Cassandra "Column family ID mismatch" into the CassandraClientPool causing the AtlasDB Cassandra connections to close.  This is symptomatic of a larger issue which is that the creation of internal tables is still unprotected and continues to suffer from collisions when multiple nodes are started (near) simultaneously.

The stack trace below demonstrates a duel submission for the _locks table which results in a Column family ID mismatch which will continue to cause issues for DefsTables.java#mergeColumnFamilies in Cassandra which is run after receiving schema mutations announcements.

Cassandra.log:
```
cassandra_1 | INFO  13:21:18 Create new ColumnFamily: org.apache.cassandra.config.CFMetaData@22e774ea[cfId=938f2540-190d-11e6-94a6-1d00381e37f2,ksName=<redacted>,cfName=_metadata,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.BytesType,org.apache.cassandra.db.marshal.LongType),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=345600,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column2, type=org.apache.cassandra.db.marshal.LongType, kind=CLUSTERING_COLUMN, componentIndex=1, indexName=null, indexType=null}, ColumnDefinition{name=value, type=org.apache.cassandra.db.marshal.BytesType, kind=COMPACT_VALUE, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column1, type=org.apache.cassandra.db.marshal.BytesType, kind=CLUSTERING_COLUMN, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={},compressionParameters={},bloomFilterFpChance=<null>,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=NONE,droppedColumns={},triggers=[],isDense=true]
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_keyspaces: 497 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_keyspaces@107282453(0.145KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-tmp-ka-6-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=124369)
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columnfamilies: 5924 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columnfamilies@1479462876(1.291KiB serialized bytes, 27 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-tmp-ka-2-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=124369)
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columns: 7464 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columns@1751779132(1.136KiB serialized bytes, 28 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-tmp-ka-2-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=124369)
cassandra_1 | INFO  13:21:18 Loading org.apache.cassandra.config.CFMetaData@5e0de6f8[cfId=938f2540-190d-11e6-94a6-1d00381e37f2,ksName=<redacted>,cfName=_metadata,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.BytesType,org.apache.cassandra.db.marshal.LongType),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=345600,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column2, type=org.apache.cassandra.db.marshal.LongType, kind=CLUSTERING_COLUMN, componentIndex=1, indexName=null, indexType=null}, ColumnDefinition{name=value, type=org.apache.cassandra.db.marshal.BytesType, kind=COMPACT_VALUE, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column1, type=org.apache.cassandra.db.marshal.BytesType, kind=CLUSTERING_COLUMN, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={},compressionParameters={},bloomFilterFpChance=<null>,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=NONE,droppedColumns={},triggers=[],isDense=true]
cassandra_1 | INFO  13:21:18 Initializing <redacted>._metadata
cassandra_1 | INFO  13:21:18 Create new ColumnFamily: org.apache.cassandra.config.CFMetaData@6d4ae6f0[cfId=93a34980-190d-11e6-94a6-1d00381e37f2,ksName=<redacted>,cfName=_locks,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.BytesType,org.apache.cassandra.db.marshal.LongType),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=345600,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column2, type=org.apache.cassandra.db.marshal.LongType, kind=CLUSTERING_COLUMN, componentIndex=1, indexName=null, indexType=null}, ColumnDefinition{name=value, type=org.apache.cassandra.db.marshal.BytesType, kind=COMPACT_VALUE, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column1, type=org.apache.cassandra.db.marshal.BytesType, kind=CLUSTERING_COLUMN, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={},compressionParameters={},bloomFilterFpChance=<null>,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=NONE,droppedColumns={},triggers=[],isDense=true]
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_keyspaces: 497 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_keyspaces@1607205712(0.145KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-tmp-ka-7-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=127908)
cassandra_1 | INFO  13:21:18 Create new ColumnFamily: org.apache.cassandra.config.CFMetaData@115175d3[cfId=93a6f300-190d-11e6-94a6-1d00381e37f2,ksName=<redacted>,cfName=_locks,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.BytesType,org.apache.cassandra.db.marshal.LongType),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=345600,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column2, type=org.apache.cassandra.db.marshal.LongType, kind=CLUSTERING_COLUMN, componentIndex=1, indexName=null, indexType=null}, ColumnDefinition{name=value, type=org.apache.cassandra.db.marshal.BytesType, kind=COMPACT_VALUE, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column1, type=org.apache.cassandra.db.marshal.BytesType, kind=CLUSTERING_COLUMN, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={},compressionParameters={},bloomFilterFpChance=<null>,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=NONE,droppedColumns={},triggers=[],isDense=true]
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columnfamilies: 5843 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columnfamilies@806976264(1.212KiB serialized bytes, 27 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-tmp-ka-3-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=127908)
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columns: 7380 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columns@1271408904(1.054KiB serialized bytes, 28 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-tmp-ka-3-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=127908)
cassandra_1 | INFO  13:21:18 Loading org.apache.cassandra.config.CFMetaData@7182fc16[cfId=93a34980-190d-11e6-94a6-1d00381e37f2,ksName=<redacted>,cfName=_locks,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.BytesType,org.apache.cassandra.db.marshal.LongType),comment=,readRepairChance=0.0,dcLocalReadRepairChance=0.1,gcGraceSeconds=345600,defaultValidator=org.apache.cassandra.db.marshal.BytesType,keyValidator=org.apache.cassandra.db.marshal.BytesType,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=key, type=org.apache.cassandra.db.marshal.BytesType, kind=PARTITION_KEY, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column2, type=org.apache.cassandra.db.marshal.LongType, kind=CLUSTERING_COLUMN, componentIndex=1, indexName=null, indexType=null}, ColumnDefinition{name=value, type=org.apache.cassandra.db.marshal.BytesType, kind=COMPACT_VALUE, componentIndex=null, indexName=null, indexType=null}, ColumnDefinition{name=column1, type=org.apache.cassandra.db.marshal.BytesType, kind=CLUSTERING_COLUMN, componentIndex=0, indexName=null, indexType=null}],compactionStrategyClass=class org.apache.cassandra.db.compaction.LeveledCompactionStrategy,compactionStrategyOptions={},compressionParameters={},bloomFilterFpChance=<null>,memtableFlushPeriod=0,caching={"keys":"ALL", "rows_per_partition":"NONE"},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=NONE,droppedColumns={},triggers=[],isDense=true]
cassandra_1 | INFO  13:21:18 Initializing <redacted>._locks
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_keyspaces: 497 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_keyspaces@2004330113(0.145KiB serialized bytes, 3 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-tmp-ka-8-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=131447)
cassandra_1 | INFO  13:21:18 Compacting [SSTableReader(path='/var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-ka-8-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-ka-5-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-ka-7-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-ka-6-Data.db')]
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columnfamilies: 5843 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columnfamilies@155653294(1.212KiB serialized bytes, 27 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-tmp-ka-4-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=131447)
cassandra_1 | INFO  13:21:18 Compacting [SSTableReader(path='/var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-1-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-4-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-3-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-2-Data.db')]
cassandra_1 | INFO  13:21:18 Enqueuing flush of schema_columns: 7380 (0%) on-heap, 0 (0%) off-heap
cassandra_1 | INFO  13:21:18 Compacted 4 sstables to [/var/lib/cassandra/data/system/schema_keyspaces-b0f2235744583cdb9631c43e59ce3676/system-schema_keyspaces-ka-9,].  814 bytes to 337 (~41% of original) in 35ms = 0.009183MB/s.  7 total partitions merged to 4.  Partition merge counts were {1:3, 4:1, }
cassandra_1 | INFO  13:21:18 Writing Memtable-schema_columns@1223788932(1.054KiB serialized bytes, 28 ops, 0%/0% of on/off-heap limit)
cassandra_1 | INFO  13:21:18 Completed flushing /var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-tmp-ka-4-Data.db (0.000KiB) for commitlog position ReplayPosition(segmentId=1463145671706, position=132025)
cassandra_1 | INFO  13:21:18 Compacting [SSTableReader(path='/var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-ka-1-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-ka-4-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-ka-2-Data.db'), SSTableReader(path='/var/lib/cassandra/data/system/schema_columns-296e9c049bec3085827dc17d3df2122a/system-schema_columns-ka-3-Data.db')]
cassandra_1 | INFO  13:21:18 Compacted 4 sstables to [/var/lib/cassandra/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-5,].  9,305 bytes to 7,817 (~84% of original) in 64ms = 0.116482MB/s.  5 total partitions merged to 3.  Partition merge counts were {1:2, 3:1, }
cassandra_1 | ERROR 13:21:18 Exception in thread Thread[MigrationStage:1,5,main]
cassandra_1 | java.lang.RuntimeException: org.apache.cassandra.exceptions.ConfigurationException: Column family ID mismatch (found 93a6f300-190d-11e6-94a6-1d00381e37f2; expected 93a34980-190d-11e6-94a6-1d00381e37f2)
```

internal_product.log:
```
dev_1 | ERROR [2016-05-13 13:21:18,105] com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPool: Startup checks failed, was not able to create the keyspace or ensure it already existed.
dev_1 | Exception in thread "main" java.lang.RuntimeException: org.apache.thrift.TException: No host tried was able to create the keyspace requested.
```

**Why was this not detected sooner:**

We believe this issue has gone unnoticed for sometime since it is rare for HA testing to exist for our current users and rare to be used in the field currently.  

It is also worth noting that if this issue is encountered in the field it would be enough to restart the crashed AtlasDB clients (might even happen automatically) as a second run would by pass the offending table clash since all required tables had been created by the last run.  In this manner the collisions would go undetected until Cassandra had been restarted resulting in any colliding tables switching schema versions to a version lacking data.  

The internal product had made test modifications to bypass this issue but has had the failures occur since use with ~0.3.33.0

**Comment by jboreiko** [comment]
Open questions:
- Can we move _metadata creation to regular table creatoin?
- Is there a way to protect the creation of _locks OR can we accept the possibility of collision on this particular table?  (There is a known danger that a Cassandra restart could cause the lock to be held twice)
- In the short term should we put some enforcement in place to ensure that an init process has been run before starting a new cluster.  This solution would like like initializationCli.sh creates the internal tables as a single node after which point AtlasDB can be started freely.  If the internal tables have not been created AtlasDB refuses to start.

**Comment by tpetracca** [comment]
"In the short term should we put some enforcement in place to ensure that an init process has been run before starting a new cluster. This solution would like like initializationCli.sh creates the internal tables as a single node after which point AtlasDB can be started freely. If the internal tables have not been created AtlasDB refuses to start."

This was exactly what I was going to suggest.  The safest (and quick/easy) implementation in the immediate term is the kvs expecting a _locks table to exist and failing if it does not, plain and simple.  It is the job of a one-off initialization cli to be run offline to populate the _locks table.

Also in the immediate term we should protect creation of all tables including internal ones like _metadata with the lock

**Comment by tpetracca** [comment]
For long term, I'm still not convinced why leadership election / paxos doesn't get us what we want here.  @gbrova mentions this in the previous issue when brought up.  I don't think HA is a problem we should be thinking about when considering _locks creation because it is a one-time event that is only an issue the very first time we try an use an atlas context.  From that point on the existence of the table (and our usage of it) protects us from future schema changes and is inherently HA.  Can we just use paxos to determine a leader, and make that person in charge of creating the _locks table.  We can make them permanently in charge with the failure scenario being if they become unresponsive between the time they're elected leader and the time when they call create table for _locks on the cassandra client we either just reboot them or run a cli to invalidate paxos state on all nodes.

**Comment by tpetracca** [comment]
Again I realize my suggestion involves an inherently non-HA thing on start-up but this is a 1-time job so I'm not convinced it matters.  The only benefit of the latter method over the "use a cli to init" is that the _locks table creation becomes a part of the startup and thus for ~99% of services deployed the functionality will work that same as it did previously (i.e. I didn't have to run a special init cli)

**Comment by clockfort** [comment]
I also don't want to rely on external init.

We can accept _locks collision. The situation you're talking about does exist, you'd need to reboot a quorum of nodes, not just one, else the lock CASs would fail. We can afford to have dispatch up anyways with a quorum of nodes missing.

We can move _metadata to happen very slightly later, it'll just be annoying.

**Comment by gbrova** [comment]
@tpetracca I probably misunderstood you last night, because I agree with everything you said here. The solution you described _will_ be HA after the _locks table gets created, but the (one-time) creation is not. 

A couple of other thoughts: 

- If we're comfortable with the one-time job not being HA, a much simpler implementation could be to just pick a "leader" at random instead of using paxos/election. (The tradeoff would be a slightly higher failure probability, but maybe we don't care about that if the cost of failure is just "you have to restart atlas")
- If we do want everything to be HA, we might consider creating multiple _locks tables and then running a consensus algorithm to decide which one to use. For example, the initialization protocol could be that every node creates a _locks_nodeid table and then participates in a single round of paxos to agree on which one the "real" _locks table is.

**Comment by jboreiko** [comment]
@clockfort I do not consider this issue to be closed.  I agree that moving _metadata table creation as per #447 is a step in the right direction but there is still a possible path to silent corruption.  I agree that the issue is now lower priority given the situation is extraordinary but I would like to have the end result of this ticket give us assurance that corruption will not happen.

**Comment by joelea** [comment]
Going to solve this via the mechanism outlined in #478

**Comment by joelea** [comment]
Waiting on conclusion of discussion #479

**Comment by joelea** [comment]
We're going to do this based on election of a table.

**Comment by clockfort** [comment]
Personally I'm okay with the _locks table exhibiting the problem, as the only bad case is if someone decides to reboot a quorum of nodes while cassandra has atlas open locks to create tables (and have those nodes unluckily be the losers in a contest of which locks Cf ID is considered canonical), which will already force table creates to screech to a halt, as these require all nodes to be up.

This minutia, for something that exists for a few seconds on the first stack bootup, needing two rounds of paxos (one internal to us, another internal to Cassandra) and locks that are more persisted and HA than our actual locks for normal stack usage, seems like massive overkill.

**Comment by markelliot** [comment]
Sorry if I'm missing something, but #479 doesn't seem to document how this will be solved, where might I discover the expected approach to solving this?

**Comment by gsheasby** [comment]
The approach we're taking is to have each node create a _locks_<UUID> table, and then have a round of Paxos to reach consensus on which is the "real" locks table. The selected table will then be marked as "real", and the other tables will be deleted.

**Comment by rhero** [comment]
Given our other current priorities and the [low probability](https://github.com/palantir/atlasdb/issues/444#issuecomment-221612886) of this being an issue going forward, we're opting for a simpler solution where we will allow users to specify a "distinguished leader" that will be used to create the initial _locks table on first-time startup.
